source/common/http/filter/lua/lua_filter.h

```diff
@@ -0,0 +1,310 @@
+
+  FilterHeadersStatus start(int function_ref);
+  FilterDataStatus onData(Buffer::Instance& data, bool end_stream);
+  FilterTrailersStatus onTrailers(HeaderMap& trailers);
+
+  void onReset() {
+    if (http_request_) {
+      http_request_->cancel();
+      http_request_ = nullptr;
+    }
+  }
+
+  static ExportedFunctions exportedFunctions() {
+    return {{"headers", static_luaHeaders},       {"body", static_luaBody},
+            {"bodyChunks", static_luaBodyChunks}, {"trailers", static_luaTrailers},
+            {"logTrace", static_luaLogTrace},     {"logDebug", static_luaLogDebug},
+            {"logInfo", static_luaLogInfo},       {"logWarn", static_luaLogWarn},
+            {"logErr", static_luaLogErr},         {"logCritical", static_luaLogCritical},
+            {"httpCall", static_luaHttpCall},     {"respond", static_luaRespond}};
+  }
+
+private:
+  /**
+   * Perform an HTTP call to an upstream host.
+   * @param 1 (string): The name of the upstream cluster to call. This cluster must be configured.
+   * @param 2 (table): A table of HTTP headers. :method, :path, and :authority must be defined.
+   * @param 3 (string): Body. Can be nil.
+   * @param 4 (int): Timeout in milliseconds for the call.
+   * @return headers (table), body (string/nil)
+   */
+  DECLARE_LUA_FUNCTION(StreamHandleWrapper, luaHttpCall);
+
+  /**
+   * Perform an inline response. This call is currently only valid on the request path. Further
+   * filter iteration will stop. No further script code will run after this call.
+   * @param 1 (table): A table of HTTP headers. :status must be defined.
+   * @param 2 (string): Body. Can be nil.
+   */
+  DECLARE_LUA_FUNCTION(StreamHandleWrapper, luaRespond);
+
+  /**
+   * @return a handle to the headers.
+   */
+  DECLARE_LUA_FUNCTION(StreamHandleWrapper, luaHeaders);
+
+  /**
+   * @return a handle to the full body or nil if there is no body. This call will cause the script
+   *         to yield until the entire body is received (or if there is no body will return nil
+   *         right away).
+   *         NOTE: This call causes Envoy to buffer the body. The max buffer size is configured
+   *         based on the currently active flow control settings.
+   */
+  DECLARE_LUA_FUNCTION(StreamHandleWrapper, luaBody);
+
+  /**
+   * @return an iterator that allows the script to iterate through all body chunks as they are
+   *         received. The iterator will yield between body chunks. Envoy *will not* buffer
+   *         the body chunks in this case, but the script can look at them as they go by.
+   */
+  DECLARE_LUA_FUNCTION(StreamHandleWrapper, luaBodyChunks);
+
+  /**
+   * @return a handle to the trailers or nil if there are no trailers. This call will cause the
+   *         script to yield of Envoy does not yet know if there are trailers or not.
+   */
+  DECLARE_LUA_FUNCTION(StreamHandleWrapper, luaTrailers);
+
+  /**
+   * Log a message to the Envoy log.
+   * @param 1 (string): The log message.
```

Comment:
One day we will render docs with Doxygen and some of this might end up looking might weird, but until that day comes, sure.